API: add Platform (OS and Architecture) to /containers/json

jsternberg opened the pull request:
Adds platform information to containers (for `docker ps`).

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Added the platform to the `container.Summary` struct so it could be returned in `/containers/json`.

**- How I did it**

Adapted this PR for the current master and upcoming api version: https://github.com/moby/moby/pull/42464

As mentioned in the original PR, there's still some follow up. I'm copying and pasting that message here.

- filter by architecture
- perhaps a way to indicate that a container/image is non-native (useful to provide that information, so that clients don't have to replicate the "variants" such as arm/v5 running on arm64 (which is not "matching", but no emulation should be needed in that case); similarly, Windows OS versions (?)
- check for consistency, e.g. docker inspect shows a Platform: linux, but does not contain Architecture.

The last one seems to be a potentially big issue with the name. The `Platform` is used instead of the name `OS`, but `Platform` in that API is the same as the OS.

**- How to verify it**

```
docker run -d --name foo nginx:alpine

curl --unix-socket /var/run/docker.sock http://localhost/containers/json | jq
```

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
`GET /containers/json` now returns an `ImageManifestDescriptor` field matching the same field in `/containers/{name}/json`. This field is only populated if the daemon provides a multi-platform image store.
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Ah, right, I need to refresh my memory as I recall there was some back and forth on doing this and I think some considered that it was part of the image info, but it's a long time since I worked on https://github.com/moby/moby/pull/42464

thaJeztah commented:
(i.e. ISTR there was some suggestion to include the image descriptor in the response or something along those lines ... but I need to dig really deep in memory 🤣)

vvoland reviewed (commented):
Hmm, wondering if we should consider doing the same as we do for `container inspect` (`/containers/.../json`), where we return the full image manifest descriptor with the `Platform` field set.

See: https://github.com/moby/moby/pull/48855

jsternberg commented:
@vvoland I'd say yes, but the `Platform` name is currently taken by the `OS` so it would be a breaking change. I'm not sure if there's a better name that can be used for it or how to resolve the breaking change.

thaJeztah commented on api/server/router/container/container_routes.go:
```diff
@@ -113,6 +113,13 @@ func (c *containerRouter) getContainersJSON(ctx context.Context, w http.Response
 		}
 	}
 
+	if versions.LessThan(version, "1.48") {
+		// ImageManifestDescriptor information was added in API 1.48
+		for _, c := range containers {
+			c.ImageManifestDescriptor = nil
+		}
+	}
+
```
For  follow-up; I'm curious what would perform better; 2 separate loops (as it's right now), or a single loop, but with a `if` in each cycle. We should only do so if it makes a real difference, as I don't think it makes the code more readable 😅 

```go
// HostConfig.Annotations because was added in API v1.46.
noAnnotations := versions.LessThan(version, "1.46")

// ImageManifestDescriptor information was added in API 1.48
noManifests := versions.LessThan(version, "1.48")

if noAnnotations || noManifests {
    for _, c := range containers {
        if noAnnotations {
            // Technically, we could even combine and this condition could also strip manifests, as < 1.46 is also < 1.48
            c.HostConfig.Annotations = nil
        }
        if noManifests {
            c.ImageManifestDescriptor = nil
        }
    }
}
```

I guess this would still work though if we want to avoid looping multiple times on API < 1.46;

```gp
if versions.LessThan(version, "1.46") {
    for _, c := range containers {
        // Ignore HostConfig.Annotations because it was added in API v1.46.
        c.HostConfig.Annotations = nil

        // ImageManifestDescriptor information was added in API 1.48
        c.ImageManifestDescriptor = nil
    }
} else if versions.LessThan(version, "1.48") {
    for _, c := range containers {
        // ImageManifestDescriptor information was added in API 1.48
        c.ImageManifestDescriptor = nil
    }
}
```

vvoland replied:
I don't think there should be a big difference between two vs single loop.

Optimistically, we should be on the latest API version, so none of these would be executed.

In the two loop case, we avoid branching inside, so that should be faster, on the other hand, these branching condition is constant, so in the single loop case, the CPU's branch predictor should take care of that.

Personally I wouldn't worry about that here, and just go with the most readable one (which is the one we have now IMO), unless we have a concrete benchmark that shows a performance difference :)

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -213,6 +213,12 @@ func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter,
 			b.VirtualSize = b.Size //nolint:staticcheck // ignore SA1019: field is deprecated, but still set on API < v1.44.
 		}
 	}
+	if versions.LessThan(version, "1.48") {
+		// ImageManifestDescriptor information was added in API 1.48
+		for _, c := range systemDiskUsage.Containers {
+			c.ImageManifestDescriptor = nil
+		}
+	}
```
Good point. ISTR that one was the horrible type that was reused "for convenience", but keeps popping up?

thaJeztah commented on api/types/container/container.go:
```diff
@@ -121,19 +121,20 @@ type State struct {
+	Names                   []string
+	Image                   string
+	ImageID                 string
+	ImageManifestDescriptor *ocispec.Descriptor `json:",omitempty"`
```
oh! Before I blame myself; we've been very lax on specifying the _intended_ casing of fields in JSON (depending on "whatever Go makes of it); this has resulted in inconsistencies where the Swagger documented the wrong casing (the `Id` one is a good example of that, where we had to later adjust the case to what it happened to be). Go itself treats JSON case-insensitive, but not all languages do (and ISTR Go is working on a "v2" JSON package to allow case-sensitivity).

Probably good to add an explicit `json:"ImageManifestDescriptor,omitempty"` (at least for new fields; we should also tackle existing fields at some point)

neersighted reviewed (approved):
LGTM with CI passing.

jsternberg committed:
API: add Platform (OS and Architecture) to /containers/json

Adds platform information to containers (for `docker ps`).

Signed-off-by: Jonathan A. Sternberg <jonathan.sternberg@docker.com>

jsternberg commented:
I've updated the test to only run when a multi-platform store is in use. I copied this from `TestInspectImageManifestPlatform` and I have also renamed the test function to match this format.

thaJeztah reviewed (approved):
LGTM

vvoland commented on container/view.go:
```diff
@@ -416,6 +417,10 @@ func (v *View) transform(ctr *Container) *Snapshot {
 	snapshot.NetworkSettings = &container.NetworkSettingsSummary{Networks: networks}
 
+	if ctr.ImageManifest != nil && ctr.ImageManifest.Platform == nil {
+		ctr.ImageManifest.Platform = &ctr.ImagePlatform
```
`ctr` is supposed to be immutable, we shouldn't patch it in `ctr` directly.

This should be patched in `snapshot.Summary`

vvoland replied:
cc @jsternberg
